examples/raja-launch-mat-mat.cpp

```diff
@@ -0,0 +1,690 @@
+ */
+
+/*
+  Define number of threads in x and y dimensions of a CUDA thread block
+*/
+#if defined(RAJA_ENABLE_CUDA)
+#define CUDA_BLOCK_SIZE 16
+#endif
+
+#if defined(__CUDA_ARCH__)
+#define TEAM_SHARED __shared__
+#define TEAM_SYNC() __syncthreads()
+#else
+#define TEAM_SHARED
+#define TEAM_SYNC() 
+#endif
+
+//
+// Define dimensionality of matrices.
+//
+const int DIM = 2;
+
+//
+// Define macros to simplify row-col indexing (non-RAJA implementations only)
+//
+// _matmult_macros_start
+#define A(r, c) A[c + N * r]
+#define B(r, c) B[c + N * r]
+#define C(r, c) C[c + N * r]
+// _matmult_macros_end
+
+/*
+  Define CUDA matrix multiplication kernel for comparison to RAJA version
+  Adapted from CUDA programming guide: https://docs.nvidia.com/cuda/cuda-c-programming-guide/
+*/
+#if defined(RAJA_ENABLE_CUDA)
+__global__ void matMultKernel(int N, double* C, double* A, double* B)
+{
+  // Block row and column
+  const int by = blockIdx.y;
+  const int bx = blockIdx.x;
+
+   // Each thread computes one element of Csub
+  // by accumulating results into Cvalue
+  double Cvalue(0.0);
+
+  // Thread row and column within local Csub
+  const int ty = threadIdx.y; //local row
+  const int tx = threadIdx.x; //local column
+
+  const int row = by * CUDA_BLOCK_SIZE + ty;  // Matrix row index
+  const int col = bx * CUDA_BLOCK_SIZE + tx;  // Matrix column index
+
+  // Shared memory used to store Asub and Bsub respectively
+  __shared__ double As[CUDA_BLOCK_SIZE][CUDA_BLOCK_SIZE];
+  __shared__ double Bs[CUDA_BLOCK_SIZE][CUDA_BLOCK_SIZE];
+
+  // Loop over all the sub-matrices of A and B that are
+  // required to compute Csub
+  // Multiply each pair of sub-matrices together
+  // and accumulate the results
+  for (int m = 0; m < (N / CUDA_BLOCK_SIZE); ++m) {
+
+    // Load Asub and Bsub from device memory to shared memory
+    // Each thread loads one element of each sub-matrix
+    As[ty][tx] = A[row*N + m*CUDA_BLOCK_SIZE + tx];
+    Bs[ty][tx] = B[(m*CUDA_BLOCK_SIZE + ty)*N + col];
+
+    // Synchronize to make sure the sub-matrices are loaded
+    // before starting the computation
+    __syncthreads();
+
+    // Multiply Asub and Bsub together
+    for (int e = 0; e < CUDA_BLOCK_SIZE; ++e)
+      Cvalue += As[ty][e] * Bs[e][tx];
+
+    // Synchronize to make sure that the preceding
+    // computation is done before loading two new
+    // sub-matrices of A and B in the next iteration
+    __syncthreads();
+  }
+
+  // Write Csub to device memory
+  // Each thread writes one element
+  C[col + N*row] = Cvalue;
+
+}
+#endif
+
+//
+// Functions for checking results
+//
+template <typename T>
+void checkResult(T *C, int N);
+
+template <typename T>
+void checkResult(RAJA::View<T, RAJA::Layout<DIM>> Cview, int N);
+
+//
+// Functions for printing results
+//
+template <typename T>
+void printResult(T *C, int N);
+
+template <typename T>
+void printResult(RAJA::View<T, RAJA::Layout<DIM>> Cview, int N);
+
+
+namespace RAJA {
```

Comment:
It looks like this and *raja-launch.cpp* share about 300 locs. Could we factor that out of both and into a shared file so this it's easier to update and review?

Reply:
Yes absolutely! @ajkunen should we merge your example to raja-launch-mat-mat.cpp  and rename to raja-teams.cpp  (we can also delete the old raja-teams.cpp)

Reply:
yeah!